cluster/node.go

```diff
@@ -75,18 +116,137 @@ func (n *Node) SetState(s NodeState) {
 	n.Unlock()
 }
 
-// provide thread safe json Marshaling
-func (n *Node) MarshalJSON() ([]byte, error) {
+func (n *Node) SetPartitions(part []int32) {
+	n.Lock()
+	n.partitions = part
+	n.Unlock()
+}
+func (n *Node) GetPartitions() []int32 {
+	n.RLock()
+	part := n.partitions
+	n.RUnlock()
+	return part
+}
+
+func (n *Node) Probe() {
+	update := nodeJS{}
+	resp, err := n.Get("/node", nil)
+	n.Lock()
+	defer n.Unlock()
+	if err != nil {
+		log.Warn("cluster: failed to get status of peer %s. %s", n.name, err)
+		if n.state != NodeUnreachable {
+			n.state = NodeUnreachable
+			n.stateChange = time.Now()
+		}
+		return
+	}
+
+	err = json.Unmarshal(resp, &update)
+	if err != nil {
+		log.Warn("cluster: failed to decode response from peer at address %s: %s\n\n%s", n.remoteAddr.String(), err, resp)
+		if n.state != NodeNotReady {
+			n.state = NodeNotReady
+			n.stateChange = time.Now()
+		}
+		return
+	}
+
+	if n.state.String() != update.State {
+		log.Info("cluster: node %s in new state: %s", update.Name, update.State)
+	}
+
+	n.name = update.Name
```

Comment:
it is ok for instance names to change.  The only requirement is that they are unique within the cluster.